File: lab4/java/SimpleSet.java
import java.util.Iterator;

/**
 * A set implemented as a SimpleMap from keys to nothing at all.
 * (It's only the keys that are used.)
 */
abstract class SimpleSet<Key> implements Iterable<Key> {
    // There are no good builtin singleton classes in Java, 
    // so we use a boolean instead:
    SimpleMap<Key, Boolean> map;

    /**
     * Returns true if there are no elements.
     */
    public boolean isEmpty() {
        return this.map.isEmpty();
    }

    /**
     * Returns true if the element is in the set.
     */
    public boolean contains(Key key) {
        return this.map.containsKey(key);
    }

    /**
     * Returns the number of elements.
     */
    public int size() {
        return this.map.size();
    }

    /**
     * Adds the given element, does nothing if it already exists.
     */
    public void add(Key key) {
        this.map.put(key, true);
    }

    /**
     * Removes all elements.
     */
    public void clear() {
        this.map.clear();
    }

    /**
     * Returns an iterator over the elements in the set.
     */
    public Iterator<Key> iterator() {
        return this.map.iterator();
    }

    /** 
     * Returns a string representation of the set.
     */
    public String toString() {
        return "SimpleSet(" + this.map + ")";
    }

    /**
     * Validates that the set is correctly implemented according to the specification.
     * @throws IllegalArgumentException if there is anything wrong.
     */
    public void validate() {
        this.map.validate();
    };

}

File: lab4/python/simple_set.py
from typing import Iterable, Iterator, Any

from simple_map import SimpleMap, Key


class SimpleSet(Iterable[Key]):
    """
    A set implemented as a SimpleMap from keys to nothing at all.
    (It's only the keys that are used.)
    """
    _map: SimpleMap[Key, Any]

    def isEmpty(self) -> bool:
        """Returns true if there are no elements."""
        return self._map.isEmpty()

    def contains(self, key: Key) -> bool:
        """Returns true if the element is in the set."""
        return self._map.containsKey(key)

    def __len__(self) -> int:
        return self.size()

    def size(self) -> int:
        """Returns the number of elements."""
        return self._map.size()

    def add(self, key: Key):
        """Adds the given element, does nothing if it already exists."""
        self._map.put(key, ...)

    def clear(self):
        """Removes all elements."""
        self._map.clear()

    def __iter__(self) -> Iterator[Key]:
        """Returns an iterator over the elements in the set."""
        return iter(self._map)

    def __str__(self) -> str:
        """Returns a string representation of the set."""
        return f"SimpleSet({self._map})"

    def validate(self):
        """
        Validates that the set is correctly implemented according to the specification.
        Raises a ValueError if there is anything wrong.
        """
        self._map.validate()


